### ISSUE TITLE
Investigate problems with some value types not using the static methods

### ISSUE THREAD

**Issue by quintesse** [ceylon/ceylon-compiler#1560]
Things like

```
Object obj  = false;
print(false == obj);
print(false == false of Object);
print(false.hash);
```

are all using boxing instead of the static value type methods.

**Comment by tombentley** [comment]
Following my change for #1996 this seems to be the only remaining place where we're sometimes using boxed instance methods where we could be using value methods (or better). I say "or better" because we know at compile time that something like `2=='c'` must be false. Do we want to optimize `==` that aggressively?

**Comment by quintesse** [comment]
> Do we want to optimize == that aggressively?

Personally I don't think it's necessary

**Comment by tombentley** [comment]
I basically agree but in such cases we have to box the RHS in order to use even the value method, so the cost is higher than you might think.

**Comment by quintesse** [comment]
Box the RHS? Why? But if you're right then it's another thing of course :) then I'd optimize it.

**Comment by tombentley** [comment]
Because on Integer, for example, although we have `equals(int, Object)` we
don't have `equals(int, char)`. We could add those of course, but that's a
bit silly, in my opinion.

**Comment by quintesse** [comment]
Ah ok, of course, `equals()` has `Object` as argument. Having specialized versions isn't *that* bad an idea, but I wouldn't make it specific to equals, so that's for the future perhaps (and the idea I have wouldn't work in this case anyway). So in the mean time we could do this. (low prio)

**Comment by gavinking** [comment]
We do optimize this, right?

```ceylon
Integer x = ... ; Integer y = ... ;
Boolean bool = x==y;

**Comment by tombentley** [comment]
Yes @gavinking that one is completely unboxed using jvm `==` (it becomes fully boxed if we don't know the RHS type is `Integer`, we should be using the value method in that case).

**Comment by gavinking** [comment]
Good, what I figured.
